### DISCUSSION TITLE
Packaging Luarocks

### DISCUSSION THREAD

**Discussion by gwerbin** [Homebrew/discussions#2224] [Writing Formulae/Casks]
I am looking into packaging [Luaposix](https://luarocks.org/modules/gvvaughan/luaposix) and the [Lua bindings for Readline](https://luarocks.org/modules/peterbillam/readline). These are Lua packages that are typically distributed and installed using Luarocks.

I want to package these because Readline is an optional dependency of [Fennel](https://github.com/Homebrew/homebrew-core/blob/master/Formula/fennel.rb).

I wasn't able to find a single example of a Lua/Luarocks library in the Homebrew Core repo! I see some instances of Lua dependencies being downloaded as "resources", but not packaged as standalone Formulas.

What is the recommended procedure for packaging Luarocks packages?

I got as far as this:

```ruby
  def install
    lua = Formula["lua"]
    system "luarocks", "--lua-dir=#{lua.prefix}", \
      "make", "--deps-mode" "none", "--no-install", \
      "luaposix-35.1-1.rockspec"
  end
```

This successfully builds the C extensions in-place and saves them to `buildpath/"macos"`, but I'm not sure what to do next.

**Comment by SMillerDev** [comment]
Well, where do you want to install it?

**Reply by gwerbin** [reply]
Lua source files can go in `/usr/local/share/lua/5.4/` or `/usr/local/lib/lua/5.4/`. Shared libraries can go in `/usr/local/lib/lua/5.4/`.

I didn't think it was possible to do something like `Formula["lua"].lib.install`, is it?

I noticed [`Pathname#env_script_all_files`](https://rubydoc.brew.sh/Pathname.html#env_script_all_files-instance_method) which seems to bundle up a bunch of dependencies in a single script, but that isn't useful here, either, because we want to put our library in a central location where other things can see it.

Maybe the correct solution is to add Readline to the Fennel formula itself somehow? Currently Fennel is built and installed as an ["all-in-one" single Lua](https://github.com/bakpakin/Fennel/blob/main/Makefile#L43-L47) file with no external dependencies, but that's a special situation and can't be extended to include arbitrary 3rd-party dependencies (unless you do something really ridiculous like embed a bunch of compiled shared objects directly in the Lua script).

**Comment by carlocab** [comment]
I think we used to package Lua rocks, and then decided to stop and just vendor them into the formulae that need them. I haven't really been able to dig into the reasons why, but @DomT4 might remember if they have time to chime in.

Assuming the reasons for not packaging Lua rocks no longer apply: why not just install it into its own prefix? It'll get symlinked into `$HOMEBREW_PREFIX/{share, lib}/lua/x.y` that way making them globally accessible. The downside of this is that this install will interfere with a user's own installation if they'd ever done `luarocks install` on their own previously.

Also, why can't you just vendor Fennel's dependencies into its own prefix just like with other formulae that need Lua rocks?

**Reply by gwerbin** [reply]
> It'll get symlinked into $HOMEBREW_PREFIX/{share, lib}/lua/x.y that way making them globally accessible. The downside of this is that this install will interfere with a user's own installation if they'd ever done luarocks install on their own previously.

Ah, that makes sense! It would be something like this?

```
/usr/local/Cellar/lua-readline/3.0.0/lib/lua/5.4/readline.lua
-> /usr/local/opt/lua-readline/lib/lua/5.4/readline.lua
-> /usr/local/lib/readline.lua
```

Isn't that how Python package used to be packaged? I think it's also how other package managers like APT, Pacman, DNF, Zypper, etc. do things. So I don't have a problem with it; that's kind of the point of installing a library anyway, right?

How would I do this in Homebrew language? `lib.install_symlink` and `Dir` somehow?

> Also, why can't you just vendor Fennel's dependencies into its own prefix just like with other formulae that need Lua rocks?

Fennel appears to be packaged as a single standalone Lua script. I will try to vendor in the deps by installing Readline as a resource, and then wrapping `fennel` with `env_script_all_files`, order to set `LUA_PATH` accordingly.
